Allow inserting middleware at arbitrary points in the middleware stack

Rosa opened the pull request:
This pull request is inspired by #794. I was using Grape and found myself needing this feature, so I have built a possible way of supporting `insert`, `insert_after` and `insert_before` in a similar way as Rails does.

Adding a new middleware can now be done through `use`, `insert`, `insert_after` and `insert_before`, and the operation chosen to do so is stacked together with the middleware class, args, and blocks. The stack is built using the operations in `Grape::Endpoint#build_stack` using a new class `Grape::Middleware::Stack`, based on `ActionDispatch::MiddlewareStack` with some simplifications (for example, only classes are supported for middleware, no symbols or strings, which is being deprecated in `ActionDispatch` in any case).

The commit message offers more details on the implementation itself. If anyone has a better idea of how this feature could be implemented I'd be glad to modify my changes to follow a different approach.

This is my first PR in the Grape project and I have tried to follow the same style when adding new tests for the feature and updating the `README`, but I'd be more than happy to adapt those according to feedback 😃

dblock commented on README.md:
```diff
@@ -2812,6 +2812,26 @@ class Overwriter < Grape::Middleware::Base
+
+class API < Grape::API
+  use Overwriter
+  insert_before Overwriter, CustomOverwriter, message: 'Overwrited again.'
```
Overwrited => Overwritten :)

dblock commented on README.md:
```diff
@@ -2812,6 +2812,26 @@ class Overwriter < Grape::Middleware::Base
+class API < Grape::API
+  use Overwriter
+  insert_before Overwriter, CustomOverwriter, message: 'Overwrited again.'
+  insert 0, CustomOverwriter, message: 'Overwrited definitive.'
```
Overwritten definitive doesn't sound great in English, how about "Overwrites all other middleware."?

Rosa replied:
Sounds good! I'll change also the original `Overwriter` response with a similar sentence.

dblock commented on spec/grape/dsl/middleware_spec.rb:
```diff
@@ -6,24 +6,44 @@ module MiddlewareSpec
       end
+
+      class MyMiddleware; end
+      class MyOtherMiddleware; end
```
I know `Dummy` is declared above the same way, but these are global scope so we should use `let(:dummy) { Class.new(...) }` instead, if you don't mind changing those.

Rosa replied:
You are totally right, I'll change them 😄

dblock commented:
This is excellent, code is spot on. I made minor text or other comments if you could please amend and I'll merge.

Rosa commented:
Thanks so much for the quick review and the feedback! I think I have addressed all the comments now, but feel free to add more comments if you find anything else that can be improved :)

dblock commented on lib/grape/util/stackable_values.rb:
```diff
@@ -3,16 +3,16 @@ module Util
       attr_accessor :inherited_values
       attr_reader :new_values
-      attr_reader :froozen_values
+      attr_reader :frozen_values
```
Those froozen values were just extra frozen. Thanks for spotting this ;)

![](http://media4.giphy.com/media/laypv6pCIEDCM/giphy.gif)

dblock commented:
Merged.
